web3/utils/filters.py

```diff
@@ -166,16 +161,51 @@ def format_entry(self, entry):
         return entry
 
     def set_data_filters(self, data_filter_set):
+        """(('uint256', [12345, 54321]), ('string', ('a-single-string',)))
+        """
         self.data_filter_set = data_filter_set
         if any(data_filter_set):
-            self.data_filter_set_regex = construct_data_filter_regex(
-                data_filter_set,
-            )
+            self.data_filter_set_function = match_fn(data_filter_set)
 
     def is_valid_entry(self, entry):
-        if not self.data_filter_set_regex:
+        if not self.data_filter_set:
             return True
-        return bool(self.data_filter_set_regex.match(entry['data']))
+        return bool(self.data_filter_set_function(entry['data']))
+
+
+def normalize_data_values(type_string, data_value):
+    _type = parse_type_string(type_string)
+    if _type.base == "string":
+        if _type.arrlist is not None:
+            return tuple((bytes_to_str(value) for value in data_value))
+        else:
+            return bytes_to_str(data_value)
+    return data_value
+
+
+def bytes_to_str(value):
+    return value.decode("utf-8")
+
+
+@curry
+def match_fn(match_values_and_abi, data):
```

Comment:
I think this function deserves a docstring and some comments.  There's a lot going on here and it is quite dense.